WIP: Lockscreen: use a video for animations

xjuan opened the pull request:
Instead of implementing every animations with css or code, simply use a video
with a key color as transparency.

Videos and stills goes under data/lockscreens/lockId/

https://phabricator.endlessm.com/T24541

tchx84 commented:
I see the building blocks are there, and this is looking surprisingly simple (kudos!).

But, I see some issues that can be solved by clarifying how the user interactions with the lock-screen must work. I will explain below how the interaction works, so my in-line comments will make more sense:

0. (no key available, lock never unlocked before) the user sees the lock-screen, clicking on the lock-screen does nothing (maybe in the future it can play a ACCESS DENIED sound).
1. (key is given, lock never unlocked before) the user _still_sees the lock-screen, but with a different image so the user knows that "now it can be opened". The user clicks the lock-screen, THEN the animation plays and the lock-screen disappears.
3. (key is given, lock already unlocked) the user does not sees the lock-screen nor the animation anymore. When the user opens the toolbox, the contents behind the lock screen are shown right-away.

tchx84 commented on src/lockscreen.js:
```diff
@@ -23,21 +26,42 @@ var Lockscreen = GObject.registerClass({
+            this._css_provider,
+            Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION
+        );
+        this._playbin = new Playbin({ visible: true });
```
Keep in mind that the lock-screen image or animation, should _only_ be shown if the lock screen has never been unlocked before.

tchx84 replied:
Making it visible here is messing with some toolboxes logic where some panels/lock should only shown after another has been unlocked.

tchx84 commented on src/lockscreen.js:
```diff
@@ -23,21 +26,42 @@ var Lockscreen = GObject.registerClass({
         this._lockChangedId = 0;
         this.connect('overlay-clicked', this._onClicked.bind(this));
+
+        GLib.timeout_add_seconds(GLib.PRIORITY_LOW, 1, () => {
```
Don't forget to remove this hack!

tchx84 commented on src/lockscreen.js:
```diff
@@ -86,12 +110,34 @@ var Lockscreen = GObject.registerClass({
         if (!this._key)
             return;
         if (this._manager.hasKey(this._key)) {
+            this._style.remove_class('locked');
```
What needs to happen here is that the lock-screen should update the image to another image that tells the user that "know it can be opened".

tchx84 replied:
To be more clear, we are missing one asset per lock-screen now.

tchx84 commented on src/lockscreen.js:
```diff
@@ -86,12 +110,34 @@ var Lockscreen = GObject.registerClass({
+        return 'default';
     }
 
     _updateLockStateWithLock() {
```
I would simply move 

```
        if (!this._lock)
            return;
```

To here so we don't need to repeat it.

tchx84 commented on src/lockscreen.js:
```diff
@@ -86,12 +110,34 @@ var Lockscreen = GObject.registerClass({
+    background: url('file:///app/share/lockscreens/${dir}/locked.png') no-repeat center;
+}`
+                );
+        } catch (e) {
```
Just curious, can this really happen?

tchx84 commented on src/lockscreen.js:
```diff
@@ -86,12 +110,34 @@ var Lockscreen = GObject.registerClass({
+
         if (!this._lock)
             return;
         if (!this._manager.isUnlocked(this._lock))
```
What needs to happen here is that:

a)  if the lock-screen `_lock` says its "locked" but the GSS says is now unlocked, then it means that the lock _just_ got unlock, therefore, the animation should be played.
 or

b) if the lock-screen `_lock` says its already unlock, and the GSS agrees, then it means it was already unlock, therefore we shouldn't even dislay the lock-screen anymore, neither the animation.

tchx84 commented on src/lockscreen.js:
```diff
@@ -116,21 +162,14 @@ var Lockscreen = GObject.registerClass({
-                this.remove(this._overlay);
-                return GLib.SOURCE_REMOVE;
-            });
+            this._playbin.play();
```
Keep in mind that, the animation should only be displayed under the specific condition where 1) the lock-screen was being displayed as locked, 2) the user by clicking or maybe the clubhouse by interacting with the GSS, has update the lock GSS state to be unlocked.

So we shouldn't be playing or stopping the animation here.

tchx84 commented on src/playbin.js:
```diff
@@ -0,0 +1,70 @@
+
+        /* TODO: calculate right autocrop values */
+        this._videobox.top = 38;
+        this._videobox.bottom = 38;
```
Does this explain (?)

![screenshot from 2018-11-23 10-06-29 1](https://user-images.githubusercontent.com/481418/48945018-a3664c00-ef07-11e8-8644-aac435fb31bc.png)

tchx84 commented:
One last comment, run the lint tool!

Commit:
Lockscreen: use videos for lockscreen animations

Instead of implementing every animation with css or code, simply use a video
with green as the transparency target color.

Videos and stills goes under data/lockscreens/lockId/

https://phabricator.endlessm.com/T24541

tchx84 commented:
Looking great, some comments/questions:

Non-critical:
* Is that (see image below) the intended (for this PR) lockscreen image ?
* Remove prints!
* Is the fizzics.2 lock assets missing?
* We need to add symlinks for encyclopedia.en as well (rebase on master, and you will note that we support it now).
* See Anitas email regarding moving some locks assets around.

Critical:
* There is a strange regression with the lvl2 lock from the framework toolbox: sometimes the animation won't play, and then (more importantly) the lockscreen seems to disappear visually, but the contents can't be accessed, e.g you can't click/select the panel options. To reproduce it start the dinosaurs toolbox with the following state:

```
{
  "item.key.framework.1" : {
    "consume_after_use" : true,
    "used" : false
  },
  "item.key.framework.2" : {
    "consume_after_use" : true,
    "used" : false
  },
  "item.key.framework.3" : {
    "consume_after_use" : true,
    "used" : false
  }
}
```
 
![screenshot from 2018-11-30 08-33-38 1](https://user-images.githubusercontent.com/481418/49288518-b5a03700-f47f-11e8-8d04-276149962b52.png)

cosimoc commented on src/com.endlessm.HackToolbox.src.gresource.xml:
```diff
@@ -33,6 +33,7 @@
     <file>OperatingSystemApp/oscursormodel.js</file>
     <file>OperatingSystemApp/oswobblymodel.js</file>
     <file>OperatingSystemApp/toolbox.js</file>
+    <file>playbin.js</file>
```
Alphabetical order

cosimoc commented on src/playbin.js:
```diff
@@ -0,0 +1,157 @@
+
+    remove_class (klass) {
+        this._style.remove_class(klass);
+    }
```
I don't find it worth it to have these special methods here, since the playbin is a widget itself. Clients can simply call `get_style_context()` on it directly as they would on other widgets.

cosimoc commented on src/playbin.js:
```diff
@@ -0,0 +1,157 @@
+    _init(props) {
+        super._init(props);
+
+        this._style = this.get_style_context();
```
There's no need to cache this as a private member in JS, since GTK already does that for us.

cosimoc commented on src/playbin.js:
```diff
@@ -0,0 +1,157 @@
+            return;
+
+        const crop_x = (this._video_width - width) / 2;
+        const crop_y = (this._video_height - height) / 2;
```
Is it worth it here to check whether the crop_x/crop_y values are already the same set on the videocrop and bail our early, or does videocrop already optimize that?

cosimoc replied:
It already optimizes that.

cosimoc commented on src/playbin.js:
```diff
@@ -0,0 +1,157 @@
+            }
+        }
+
+        return GLib.SOURCE_CONTINUE;
```
Nitpick: `GstBusFunc` is defined to return `true` or `false`

cosimoc commented on src/playbin.js:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    stop () {
+        this._playbin.set_state(Gst.State.NULL);
```
Probably worth it here to check if the playbin exists.

cosimoc commented on data/extra-install.sh:
```diff
@@ -0,0 +1,14 @@
+	com.endlessm.encyclopedia.en
+"
+
+LOCKSCREENSDIR=/app/share/lockscreens
```
It would be cool to respect the build prefix here.

Also I am not sure that this script is necessary to be honest, or the best way to solve this problem. The toolbox should already know that a framework app uses lockscreens called "framework", and I'd rather have the toolbox look for that asset directly rather than adding all these symlinks here.

@ptomato perhaps you can help suggesting how to do that.

ptomato replied:
To respect the prefix in this script you can look in the environment variable `MESON_INSTALL_PREFIX`, it will be set when running in the script.

I think it would be difficult to special-case the lockscreen names. It's already determined that the locks are named per-app, since the key has to open each app's lock separately. You could name the assets according to the key IDs rather than the lock IDs, but then you would not be able to have a separate asset for the hackdex lock (because that is also opened by the framework key.) Basically I think there will have to be annoying architecture somewhere, and it might as well be here, so that we can keep the lockscreen asset names equal to the lock IDs.

cosimoc commented on src/lockscreen.js:
```diff
@@ -1,51 +1,60 @@
 
     set locked(value) {
-        if ('_locked' in this && this._locked === value)
+        if (this._locked === value)
```
I'd like to preserve the original `if('$prop' in this)` pattern here, unless there's a reason not to

cosimoc commented on src/lockscreen.js:
```diff
@@ -72,34 +81,49 @@ var Lockscreen = GObject.registerClass({
         this._lock = lock;
+
+        var path = `/app/share/lockscreens/${lock}`;
+        var dir = Gio.File.new_for_path(path);
```
I think these could be `const`?

ptomato replied:
`eslint --fix` should fix that

cosimoc commented on src/lockscreen.js:
```diff
@@ -72,34 +81,49 @@ var Lockscreen = GObject.registerClass({
             `changed::${lock}`, this._updateLockStateWithLock.bind(this));
         this._lock = lock;
+
+        var path = `/app/share/lockscreens/${lock}`;
```
It would be great to respect the build prefix, or do some kind of relative lookup.

cosimoc commented on src/lockscreen.js:
```diff
@@ -1,51 +1,60 @@
             expand: true,
-            visible: true,
+            noShowAll: true,
+            visible: false,
```
This is the default

cosimoc commented on src/lockscreen.js:
```diff
@@ -72,34 +81,49 @@ var Lockscreen = GObject.registerClass({
-            /* TODO update the lock-screen graphics with can-be-opened image */
-        }
+        if (this._assetsPath && this._manager.hasKey(this._key))
+            this._playbin.setBackground(`file://${this._assetsPath}/has-key`);
```
Should we also set the `no-key` state here in the `else` case?

cosimoc commented on src/lockscreen.js:
```diff
@@ -72,34 +81,49 @@ var Lockscreen = GObject.registerClass({
+        if (this._assetsPath) {
+            var bg = this._manager.hasKey(this._key) ? 'has-key' : 'no-key';
+            this._playbin.setBackground(`file://${this._assetsPath}/${bg}`);
+        }
```
Perhaps we can call `_updateLockStateWithKey()` here? It's unclear to me if we need to do it here in addition to what we do in `_updateLockStateWithKey()`.

cosimoc replied:
I'd also try to centralize every UI state change in `_updateUI()`, it's difficult to follow all the flow otherwise.
